### PR TITLE
Bioin 1767 ugbio mrd

### PR DESCRIPTION
**Pull request by avigailmo**

### CODE REVIEW

File: src/mrd/tests/resources/pancan_pcawg_2020.mutations_hg38_GNOMAD_dbsnp_beds.sorted.Annotated.HMER_LEN.edited.chr19.vcf.gz
Comment (inbalzelig-ugi) [inline src/mrd/tests/resources/pancan_pcawg_2020.mutations_hg38_GNOMAD_dbsnp_beds.sorted.Annotated.HMER_LEN.edited.chr19.vcf.gz:1]: This file is 7M. I think it is too big to be part of unit testing. If you agree, we should ask Itai to look if this whole file really necessary

File: src/srsnv/ugbio_srsnv/srsnv_training_utils.py [hunk 1]
```diff
@@ -0,0 +1,1314 @@
...
+            phred_residual_snv_rate[-1],
+        ),  # below ML_QUAL=1E-10 assign 0, above max assign max
+    )
+    return quality_interpolation_function
+
+
+# pylint:disable=too-many-arguments
+# pylint:disable=too-many-branches
+# pylint:disable=too-many-statements
+def prepare_featuremap_for_model(
```
Comment (inbalzelig-ugi) [inline src/srsnv/ugbio_srsnv/srsnv_training_utils.py:324]: wow this function is terrbile... It should ne splited to smaller functios, but this looks like a lot of work and not mandatory for now

File: src/srsnv/ugbio_srsnv/srsnv_training.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+# DESCRIPTION
+#    Run single read SNV quality recalibration training
+# CHANGELOG in reverse chronological order
+from __future__ import annotations
```
Comment (inbalzelig-ugi) [inline src/srsnv/ugbio_srsnv/srsnv_training.py:18]: remove

File: src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py [hunk 3]
```diff
@@ -0,0 +1,3288 @@
+from __future__ import annotations
```
Comment (inbalzelig-ugi) [inline src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py:1]: remove

File: src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py [hunk 4]
```diff
@@ -0,0 +1,3288 @@
...
+        )
+
+    logger.info(f"min_LoD_filter {best_LoD_filter} (LoD={best_lod:.1e})")
+    lod_stats = pd.Series(lod_stats_dict)
+    lod_stats.to_hdf(output_h5, key="lod_stats", mode="a")
+
+    return best_LoD_filter
+
+
+class SRSNVReport:
```
Comment (inbalzelig-ugi) [inline src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py:1486]: Maybe this class will be on a seperate file? then this one is not so huge

File: src/srsnv/ugbio_srsnv/srsnv_inference_utils.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
+from __future__ import annotations
```
Comment (inbalzelig-ugi) [inline src/srsnv/ugbio_srsnv/srsnv_inference_utils.py:1]: remove

File: src/srsnv/ugbio_srsnv/srsnv_inference.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+# DESCRIPTION
+#    Add ml qual to SNVs according to features in featuremap
+# CHANGELOG in reverse chronological order
+from __future__ import annotations
```
Comment (inbalzelig-ugi) [inline src/srsnv/ugbio_srsnv/srsnv_inference.py:18]: remove

File: src/mrd/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    cd .. && \
+    rm -rf bedops
+
+ARG OUT_DIR=/tmp/ugbio
+
+COPY --from=build $OUT_DIR $OUT_DIR
+
+RUN pip install ${OUT_DIR}/*.whl
+
+COPY --from=build ./src/mrd/ugbio_mrd/reports ./src/mrd/ugbio_mrd/reports
```
Comment (inbalzelig-ugi) [inline src/mrd/Dockerfile:100]: why you need to copy reports folder?

File: src/mrd/ugbio_mrd/prepare_data_from_mrd_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+# DESCRIPTION
+#    MRD data preparation pipeline
+# CHANGELOG in reverse chronological order
+from __future__ import annotations
```
Comment (inbalzelig-ugi) [inline src/mrd/ugbio_mrd/prepare_data_from_mrd_pipeline.py:18]: remove

File: src/mrd/ugbio_mrd/mrd_utils.py [hunk 9]
```diff
@@ -0,0 +1,810 @@
+from __future__ import annotations
```
Comment (inbalzelig-ugi) [inline src/mrd/ugbio_mrd/mrd_utils.py:1]: remove

File: src/core/ugbio_core/intersect_bed_regions.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import argparse
```
[the remark this answers is not in the document]
Comment (kenissur) [reply src/core/ugbio_core/intersect_bed_regions.py:1]: I don't understand why you are saying it should be on ugvc, it's part of the pipeline that was moved to ugbio

File: src/srsnv/pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    "scikit-learn>=1.5.1",
+    "numba==0.60.0",
+    "shap>=0.46.0",
+]
+description = "Ultima Genomics SRSNV utils"
+authors = [
+    { name = "Itai Rusinek", email = "itai.rusinek@ultimagen.com" },
+    { name = "Gat Krieger", email = "gat.krieger@ultimagen.com" },
+    { name = "Inbal Zelig", email = "inbal.zelig@ultimagen.com" },
+    { name = "Ori Hirschberg", email = "ori.hirschberg@ultimagen.com" },
```
Comment (kenissur) [inline src/srsnv/pyproject.toml:24]: +Rina, Doron

**Review by kenissur — APPROVED** [review]

File: src/core/ugbio_core/annotate_contig.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
+#!/bin/env python
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/core/ugbio_core/annotate_contig.py:1]: it's needed for being able to run from ugvc

File: src/ppmseq/ugbio_ppmseq/ppmSeq_utils.py [hunk 13]
```diff
@@ -0,0 +1,2016 @@
...
+    Convert a 1D dataframe to a 1D dataframe that is readable by Papyrus
+    """
+    df = pd.read_hdf(h5_file, key)
+    df.index.names = [""]
+    df = df.T
+    df.reset_index(drop=True, inplace=True)
+    return df
+
+
+def convert_h5_to_papyrus_json(h5_file: str, output_json: str) -> str:
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/ppmseq/ugbio_ppmseq/ppmSeq_utils.py:1735]: @kenissur ?

File: src/ppmseq/ugbio_ppmseq/ppmSeq_utils.py [hunk 14]
```diff
@@ -0,0 +1,2016 @@
...
+        # convert to html
+        subprocess.check_call(f"jupyter nbconvert {output_report_ipynb} --to html --no-input".split())
+        # edit html for readability
+        modify_jupyter_notebook_html(output_report_html)
+
+    # remove temporary png and ipynb files
+    if not keep_temp_visualization_files:
+        for visualization_file in output_visualization_files:
+            if os.path.isfile(visualization_file):
+                os.remove(visualization_file)
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/ppmseq/ugbio_ppmseq/ppmSeq_utils.py:2016]: @kenissur who can take it?

File: src/featuremap/ugbio_featuremap/featuremap_utils.py [hunk 15]
```diff
@@ -0,0 +1,796 @@
+from __future__ import annotations
+
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/featuremap/ugbio_featuremap/featuremap_utils.py:2]: again a task @kenissur should take

[the remark this answers is not in the document]
Comment (kenissur) [reply src/featuremap/ugbio_featuremap/featuremap_utils.py:2]: I don't really agree or maybe I just don't understand the comment, but in any case I have no attention for refactoring this code, I suggest we go ahead with this as is

File: src/core/ugbio_core/dna_format.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
+from pandas.api.types import CategoricalDtype
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/core/ugbio_core/dna_format.py:1]: 👍

File: src/core/ugbio_core/intersect_bed_regions.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/core/ugbio_core/intersect_bed_regions.py:1]: yes, it's used here: https://github.com/Ultimagen/terra_pipeline/blob/master/wdls/tasks/mrd.wdl#L484

File: src/core/ugbio_core/strand_direction.py [hunk 17]
```diff
@@ -0,0 +1,19 @@
+from enum import Enum
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/core/ugbio_core/strand_direction.py:1]: it came togehter with dna_utils and dna_format so I thoght to take all dna scripts togehter - but I'll revert this no problem

File: src/featuremap/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+RUN wget https://github.com/arq5x/bedtools2/releases/download/v${BEDTOOLS_VERSION}/bedtools-${BEDTOOLS_VERSION}.tar.gz && \
+    tar -zxvf bedtools-${BEDTOOLS_VERSION}.tar.gz && \
+    cd bedtools2 && \
+    make && \
+    make install && \
+    cd .. && \
+    rm -rf bedtools-${BEDTOOLS_VERSION}.tar.gz bedtools2
+
+# Install bedops
+RUN git clone https://github.com/bedops/bedops.git && \
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/featuremap/Dockerfile:74]: will do the order in dockers in separate pr

File: src/mrd/tests/resources/pancan_pcawg_2020.mutations_hg38_GNOMAD_dbsnp_beds.sorted.Annotated.HMER_LEN.edited.chr19.vcf.gz
Comment (avigailmo) [reply src/mrd/tests/resources/pancan_pcawg_2020.mutations_hg38_GNOMAD_dbsnp_beds.sorted.Annotated.HMER_LEN.edited.chr19.vcf.gz:1]: @kenissur Can you make the test file smaller?

File: src/srsnv/ugbio_srsnv/top_preprocessor.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+"""
+        toc_html = self.generate_toc_html()
+        toc_content = custom_css + custom_text + toc_html
+
+        # Insert TOC at the placeholder or at the beginning
+        self.insert_toc(nb, toc_content)
+
+        return nb, resources
+
+    def preprocess_cell(self, cell, resources, index):  # pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/srsnv/ugbio_srsnv/top_preprocessor.py:30]: I prefer Ori to do these changes

File: src/srsnv/ugbio_srsnv/srsnv_training_utils.py [hunk 20]
```diff
@@ -0,0 +1,1314 @@
...
+    return (
+        downsampled_training_featuremap_vcf,
+        downsampled_test_featuremap_vcf,
+        featuremap_entry_number,
+        train_set_size,
+        test_set_size,
+    )
+
+
+class SRSNVTrain:  # pylint: disable=too-many-instance-attributes
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/srsnv/ugbio_srsnv/srsnv_training_utils.py:611]: not in the scope of this task

File: src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py [hunk 21]
```diff
@@ -0,0 +1,3288 @@
...
+        except Exception as e:  # pylint: disable=broad-except
+            logger.error(f"Exception occurred in {func.__name__}: {e}", exc_info=True)
+            if exception_config.raise_exception:
+                raise e
+            return None
+
+    return wrapper
+
+
+def prob_to_phred(arr, eps=1e-10):
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py:91]: right good catch - we have this function also in [srsnv_training_utils](https://github.com/Ultimagen/ugbio-utils/blob/c2bd591af8db56f1a66dbe7afb272ca14ca894e1/src/srsnv/ugbio_srsnv/srsnv_training_utils.py#L191) but with different epsilon. @kenissur  Can you take it?

File: src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py [hunk 22]
```diff
@@ -0,0 +1,3288 @@
...
+
+    # calculate the discretized bins
+    left_of_first_bin = full_range[0] - float(discretization_size) / 2
+    right_of_last_bin = full_range[1] + float(discretization_size) / 2
+    discretized_bins = np.arange(left_of_first_bin, right_of_last_bin + discretized_bin_width, discretized_bin_width)
+
+    return discretized_bins
+
+
+# ### Old functions from here
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/srsnv/ugbio_srsnv/srsnv_plotting_utils.py:201]: @orihir last changes.

File: src/mrd/Dockerfile (hunk 7 above)
Comment (avigailmo) [reply src/mrd/Dockerfile:100]: so the report will be accessible from the container

File: src/mrd/ugbio_mrd/mrd_utils.py [hunk 23]
```diff
@@ -0,0 +1,810 @@
...
+        .replace("signature_", "")  # remove generic suffixes
+        .replace("_filtered", "")  # remove generic suffixes
+        .replace("filtered", "")  # remove generic suffixes
+        .replace("featuremap_", "")  # remove generic suffixes
+        .replace("_featuremap", "")  # remove generic suffixes
+        .replace("featuremap", "")  # remove generic suffixes
+    )
+
+
+def _get_hmer_length(ref: str, left_motif: str, right_motif: str):
```
[the remark this answers is not in the document]
Comment (avigailmo) [reply src/mrd/ugbio_mrd/mrd_utils.py:77]: didn't find

**Pull request merged**